### PR TITLE
Added handling for channels in different workspaces that share the same ID

### PR DESCRIPTION
**Pull request by Diana520**
I replaced channelId to composite Key - class ConversationKey that contains teamId and channelId

### CODE REVIEW

**Review by utluiz — CHANGES_REQUESTED** [review]
Most changes are good. However, the current changes aren't enough to completely migrate channel IDs to a composite key of team and channel IDs.
On top of the changes pointed here, there are many places in the code that depend on a channelId only and thus should be updated. A few of them I could find are:
- `ConfigurationDAO#findByChannel`
- `DedicatedChannelDAO#findMappingsForChannel`
- `ProjectConfigurationManager#muteProjectConfigurationsByChannelId`
- `ProjectConfigurationManager#unmuteProjectConfigurationsByChannelId`
- `ProjectConfigurationManager#getAllProjectsByChannel`
- `ProjectConfigurationManager#deleteProjectConfigurationsByChannelId`
- `DedicatedChannelManager#isNotSameChannel`
- `DefaultIssueMentionService#onChannelDeletedEvent`
- `DefaultIssueMentionService#deleteMessageMention`
- `DefaultEntityToChannelMappingManager#removeNotificationsForEntityAndChannel`
- `DefaultEntityToChannelMappingManager#removeNotificationForEntityAndChannel`
- `DefaultEntityToChannelMappingManager#removeNotificationsForChannel`
- `NotificationConfigurationDao#removeNotificationsForChannel`
- `DefaultSlackSettingsService` also contains code to mute channels by ID. The IDs are stored generally as plugins settings. Such data should be migrated to using teamId + channelId.

There could be other places but the list above is already big enough to make a really big PR. 
It could be a good idea to make such changes in parts (different PRs) as long as they can be isolated from one another.

File: confluence-slack-server-integration-plugin/src/main/java/com/atlassian/confluence/plugins/slack/spacetochannel/service/DefaultSlackSpaceToChannelService.java [hunk 1]
```diff
@@ -77,7 +77,7 @@ public DefaultSlackSpaceToChannelService(
                         .flatMap(client -> client.getConversationsInfo(channel.getChannelId()).toOptional())));
 
         return channelId -> conversationsAndLinks.linkByChannelId(channelId)
-                .map(link -> conversationsAndLinks.conversation(channelId)
+                .map(link -> conversationsAndLinks.conversation(new ConversationKey(mappings.iterator().next().getTeamId(), channelId))
```
[the remark this answers is not in the document]
Comment (Diana520) [reply confluence-slack-server-integration-plugin/src/main/java/com/atlassian/confluence/plugins/slack/spacetochannel/service/DefaultSlackSpaceToChannelService.java]: I changed String to ConversationKey and renamed variable to conversationKey

File: slack-server-integration-common/src/main/java/com/atlassian/plugins/slack/api/client/ConversationsAndLinks.java [hunk 2]
```diff
@@ -1,26 +1,27 @@
...
 import com.github.seratch.jslack.api.model.Conversation;
 
 import java.util.Map;
 import java.util.Optional;
 
 public class ConversationsAndLinks {
-    private final Map<String, Conversation> conversations;
+    private final Map<ConversationKey, Conversation> conversations;
     private final Map<String, SlackLink> links;
     private final Map<String, SlackLink> linksByChannelId;
```
[the remark this answers is not in the document]
Comment (Diana520) [reply slack-server-integration-common/src/main/java/com/atlassian/plugins/slack/api/client/ConversationsAndLinks.java]: Conflict happened and i changed it to ConversationKey

File: slack-server-integration-common/src/main/java/com/atlassian/plugins/slack/api/client/ConversationLoaderHelper.java [hunk 3]
```diff
@@ -93,38 +95,40 @@ public ConversationLoaderHelper(final SlackClientProvider slackClientProvider,
...
-                        return Optional.<Conversation>empty();
+                        return Optional.<Pair<ConversationKey, Conversation>>empty();
                     }
                 })
                 .filter(Optional::isPresent)
                 .map(Optional::get)
-                .collect(toMap(Conversation::getId, Function.identity()));
+                .collect(toMap(Pair<ConversationKey, Conversation>::getLeft, Pair<ConversationKey, Conversation>::getRight));
 
         return new ConversationsAndLinks(conversationsById, linksByTeamId, linksByChannelId);
```
[the remark this answers is not in the document]
Comment (Diana520) [reply slack-server-integration-common/src/main/java/com/atlassian/plugins/slack/api/client/ConversationLoaderHelper.java:117]: linksByChannelId now has ConversationKey type

File: jira-slack-server-integration/jira-slack-server-integration-plugin/src/main/resources/slack-soy/channelmapping/channelmapping.soy [hunk 4]
```diff
@@ -135,17 +135,20 @@
...
+                            {call JIRA.Templates.Slack.Project.ChannelMapping.channelMapping} //
+                                {param teamId: $conversationKey.teamId /}
+                                {param teamName: $projectConfiguration.channels[$convKey].teamName /}
+                                {param channelId: $conversationKey.channelId /}
+                                {param channelName: $projectConfiguration.channels[$convKey].channelName /}
                                 {param projectName: $projectConfiguration.projectName /}
                                 {param projectKey: $projectConfiguration.projectKey /}
                                 {param projectId: $projectConfiguration.projectId /}
-                                {param config: $configTeam.settings /}
+                                {param config: $configTeam.settings /} //
```
[the remark this answers is not in the document]
Comment (Diana520) [reply jira-slack-server-integration/jira-slack-server-integration-plugin/src/main/resources/slack-soy/channelmapping/channelmapping.soy]: I deleted unnecessary slashes

File: jira-slack-server-integration/jira-slack-server-integration-plugin/src/main/resources/slack-soy/channelmapping/channelmapping.soy [hunk 5]
```diff
@@ -135,17 +135,20 @@
...
-                                {param teamId: $projectConfiguration.channels[$channelId].teamId /}
-                                {param teamName: $projectConfiguration.channels[$channelId].teamName /}
-                                {param channelId: $channelId /}
-                                {param channelName: $projectConfiguration.channels[$channelId].channelName /}
+                    {foreach $conversationKey in $projectConfiguration.orderedChannelIds} // ConversationKey
+                    {let $convKey: $conversationKey.teamId + ':' + $conversationKey.channelId /}
+                        {foreach $configTeam in $projectConfiguration.configuration[$convKey]} //  List<ConfigurationGroupDTO>
+                        <!-- settings $configTeam.settings  -->
+                        <!-- config group id $configTeam.configurationGroupId -->
+                            {call JIRA.Templates.Slack.Project.ChannelMapping.channelMapping} //
```
[the remark this answers is not in the document]
Comment (Diana520) [reply jira-slack-server-integration/jira-slack-server-integration-plugin/src/main/resources/slack-soy/channelmapping/channelmapping.soy]: I needed to define the type of variables in template to find out how to change configuration class

File: slack-server-integration-common/src/main/java/com/atlassian/plugins/slack/api/client/ConversationLoaderHelper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mvlasovatl) [reply slack-server-integration-common/src/main/java/com/atlassian/plugins/slack/api/client/ConversationLoaderHelper.java:117]: It would be great to have it renamed to `linksByConversationKey` then.

File: confluence-slack-server-integration-plugin/src/main/java/com/atlassian/confluence/plugins/slack/spacetochannel/configuration/SpaceToChannelConfiguration.java [hunk 6]
```diff
@@ -117,7 +118,7 @@ public SpaceToChannelConfiguration build() {
          * @param channelId the channel id the SpaceToChannelSettings.Builder is for
          * @return a SpaceToChannelSettings.Builder
          */
-        public SpaceToChannelSettings.Builder getSettingsBuilder(String channelId) {
+        public SpaceToChannelSettings.Builder getSettingsBuilder(ConversationKey channelId) {
```
Comment (mvlasovatl) [inline confluence-slack-server-integration-plugin/src/main/java/com/atlassian/confluence/plugins/slack/spacetochannel/configuration/SpaceToChannelConfiguration.java:121]: Please rename `channelId` to `conversationKey` in order to make it easier to detect variable type by its name. It may be done in another PR.

File: slack-server-integration-common/src/main/java/com/atlassian/plugins/slack/settings/DefaultSlackSettingsService.java [hunk 7]
```diff
@@ -29,11 +30,11 @@
...
 
     @Override
-    public boolean isChannelMuted(final String channelId) {
+    public boolean isChannelMuted(final ConversationKey conversationKey) {
         boolean isMuted = false;
-        if (StringUtils.isNotBlank(channelId)) {
+        if (StringUtils.isNotBlank(conversationKey.toString())) {
             List<String> mutedChannelIds = getMutedChannelIds();
-            isMuted = mutedChannelIds.contains(channelId);
+            isMuted = mutedChannelIds.contains(conversationKey);
```
Comment (mvlasovatl) [inline slack-server-integration-common/src/main/java/com/atlassian/plugins/slack/settings/DefaultSlackSettingsService.java:37]: Here presence of `ConversationKey` value is checked in list of type `List<String>`. For some reasons `.contains()` method doesn't use generic type of the `List`, so compiler doesn't raise an error here during the compilation. Nevertheless, the result of this invocation will always be `false`. 

To address it I'd suggest to use `<teamId>:<channelId>` as a value for muting channels. It may be implemented as a new method `toStringKey()` in `ConversationKey` class. I'd like it to be addressed in this PR because it makes muting logic partially broken.

File: slack-server-integration-common/src/main/java/com/atlassian/plugins/slack/api/client/ConversationLoaderHelper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Diana520) [reply slack-server-integration-common/src/main/java/com/atlassian/plugins/slack/api/client/ConversationLoaderHelper.java:117]: It will be renamed in the next PR

File: jira-slack-server-integration/jira-slack-server-integration-plugin/src/main/java/com/atlassian/jira/plugins/slack/model/dto/ProjectToChannelConfigurationDTO.java [hunk 8]
```diff
@@ -69,7 +68,7 @@ public String getProjectKey() {
     }
 
     @SuppressWarnings("unused")
-    public List<String> getOrderedChannelIds() {
+    public List<ConversationKey> getOrderedChannelIds() {
```
Comment (mvlasovatl) [inline jira-slack-server-integration/jira-slack-server-integration-plugin/src/main/java/com/atlassian/jira/plugins/slack/model/dto/ProjectToChannelConfigurationDTO.java:71]: After changing the type list contains `getOrderedChannelIds()` method name becomes misleading. Could you rename it to `getOrderedConversationKeys()`? It may be done in another PR.

File: jira-slack-server-integration/jira-slack-server-integration-plugin/src/main/resources/slack-soy/channelmapping/channelmapping.soy [hunk 9]
```diff
@@ -135,13 +135,16 @@
                     </tr>
                   </thead>
                 {foreach $projectConfiguration in $projectConfigurations}
-                    {foreach $channelId in $projectConfiguration.orderedChannelIds}
-                        {foreach $configTeam in $projectConfiguration.configuration[$channelId]}
+                    {foreach $conversationKey in $projectConfiguration.orderedChannelIds} //ConversationKey
+                    {let $convKey: $conversationKey.teamId + ':' + $conversationKey.channelId /}
+                        {foreach $configTeam in $projectConfiguration.configuration[$convKey]} //List<ConfigurationGroupDTO>
+                        <!-- settings $configTeam.settings  -->
+                        <!-- config group id $configTeam.configurationGroupId -->
```
Comment (mvlasovatl) [inline jira-slack-server-integration/jira-slack-server-integration-plugin/src/main/resources/slack-soy/channelmapping/channelmapping.soy:142]: I suppose these comments that were used for local debugging aren't needed anymore and need to be removed.

File: slack-server-integration-common/src/main/java/com/atlassian/plugins/slack/api/client/ConversationLoaderHelper.java [hunk 10]
```diff
@@ -93,38 +95,40 @@ public ConversationLoaderHelper(final SlackClientProvider slackClientProvider,
...
 
     /**
      * Load conversation using the provided values and loader but only for the first provided item.
      */
-    private <T> Future<Optional<Conversation>> loadConversationFromList(
-            final ConversationKey channel,
-            final Map<String, SlackLink> linksByChannelId,
+    private Future<Optional<Pair<ConversationKey, Conversation>>> loadConversationFromList(
+            final ConversationKey conversationKey,
+            final Map<ConversationKey, SlackLink> linksByChannelId,
```
Comment (mvlasovatl) [inline slack-server-integration-common/src/main/java/com/atlassian/plugins/slack/api/client/ConversationLoaderHelper.java:125]: `linksByChannelId` -> `linksByConversationKey` for the sake of clarity.

**Review by mvlasovatl — CHANGES_REQUESTED** [review]

File: slack-server-integration-common/src/main/java/com/atlassian/plugins/slack/settings/DefaultSlackSettingsService.java (hunk 7 above)
Comment (utluiz) [reply slack-server-integration-common/src/main/java/com/atlassian/plugins/slack/settings/DefaultSlackSettingsService.java:37]: That's right. Yet another change that will be necessary.

**Comment by Diana520** [comment]
I finished transition from channelId to conversationKey related to Jira. Due to huge amount of replacing channelId with conversationKey, i'm going to change other classes that depend on a channelId only related to Jira, Confluence and Bitbucket and push these changes to another PR that i will create

File: bitbucket-slack-server-integration-plugin/src/main/java/com/atlassian/bitbucket/plugins/slack/listener/SlackChannelEventListener.java [hunk 11]
```diff
@@ -62,10 +67,14 @@ public void onConversationsLoadedEvent(final SlackConversationsLoadedEvent event
...
         asyncExecutor.run(() -> {
-            final List<String> mutedChannelIds = settingService.getMutedChannelIds();
+            final List<ConversationKey> mutedChannelIds = settingService.getMutedChannelIds();
             if (!mutedChannelIds.isEmpty()) {
-                Set<String> activeConversationIds = event.getConversations().stream()
-                        .map(Conversation::getId)
+                Set<ConversationKey> activeConversationIds = event.getConversations().stream()
+                        .flatMap(c -> {
+                            List<ConversationKey> list = new ArrayList<>();
+                            c.getSharedTeamIds().forEach(t -> list.add(new ConversationKey(t, c.getId())));
```
Comment (utluiz) [inline bitbucket-slack-server-integration-plugin/src/main/java/com/atlassian/bitbucket/plugins/slack/listener/SlackChannelEventListener.java:75]: Using `getSharedTeamIds` will not work as expected.
The way I think you can get the right teamId here is adding it to the SlackConversationsLoadedEvent object.
This event is created in the DefaultSlackClient class. Then, when instantiating the object there, pass `slackLink.getTeamId()`.
Finally, in this listener, you can do:
```
.map(c -> new ConversationKey(event.getTeamId(), c.getId())))
```

File: confluence-slack-server-integration-plugin/src/main/java/com/atlassian/confluence/plugins/slack/spacetochannel/listener/SlackChannelEventListener.java [hunk 12]
```diff
@@ -66,10 +70,14 @@ public void onConversationsLoadedEvent(final SlackConversationsLoadedEvent event
...
         // just in case we missed 'channel_unarchive' event
         asyncExecutor.run(() -> {
-            final List<String> mutedChannelIds = settingService.getMutedChannelIds();
+            final List<ConversationKey> mutedChannelIds = settingService.getMutedChannelIds();
             if (!mutedChannelIds.isEmpty()) {
-                Set<String> activeConversationIds = event.getConversations().stream()
-                        .map(Conversation::getId)
+                Set<ConversationKey> activeConversationIds = event.getConversations().stream()
+                        .flatMap(c -> {
+                            List<ConversationKey> list = new ArrayList<>();
```
Comment (utluiz) [inline confluence-slack-server-integration-plugin/src/main/java/com/atlassian/confluence/plugins/slack/spacetochannel/listener/SlackChannelEventListener.java:77]: Same comment as the other listener above, i.e. please add getTeamId to `SlackConversationsLoadedEvent`.

**Review by utluiz — APPROVED** [review]

**Review by mvlasovatl — APPROVED** [review]
